src/perf/counter/counter_provider.cpp

```diff
@@ -0,0 +1,194 @@
+                        << "' does not name a known event, ignoring! (reason: " << e.what() << ")";
+        }
+    }
+}
+
+void CounterProvider::initialize_group_counters(std::string leader,
+                                                std::vector<std::string> counters)
+{
+    assert(group_events_.empty());
+
+    if (leader == "")
+    {
+        Log::info() << "choosing default metric-leader";
+
+        try
+        {
+            group_leader_ = EventProvider::get_event_by_name("cpu-clock");
+        }
+        catch (const EventProvider::InvalidEvent& e)
+        {
+            Log::warn() << "cpu-clock isn't available, trying to use a fallback event";
+            try
+            {
+                group_leader_ = EventProvider::get_default_metric_leader_event();
+            }
+            catch (const perf::EventProvider::InvalidEvent& e)
+            {
+                Log::error() << "Failed to determine a suitable metric leader event";
+                Log::error() << "Try manually specifying one with --metric-leader.";
+
+                throw perf::EventProvider::InvalidEvent(leader);
+            }
+        }
+    }
+    else
+    {
+        group_leader_ = EventProvider::get_event_by_name(leader);
+    }
+
+    for (const auto& ev : counters)
+    {
+
+        try
+        {
+
+            const auto event_desc = perf::EventProvider::get_event_by_name(ev);
+
+            // skip event if it has already been declared as group leader
+            if (event_desc == group_leader_)
+            {
+                Log::info() << "'" << ev
+                            << "' has been requested as both the metric leader event and a regular "
+                               "metric event. Will treat it as the leader.";
+                continue;
+            }
+
+            group_events_.emplace_back(event_desc);
+        }
+        catch (const perf::EventProvider::InvalidEvent& e)
+        {
+            Log::warn() << "'" << ev
+                        << "' does not name a known event, ignoring! (reason: " << e.what() << ")";
+        }
+    }
+}
+
+CounterCollection CounterProvider::collection_for(MeasurementScope scope)
+{
+    assert(scope.type == MeasurementScopeType::GROUP_METRIC ||
+           scope.type == MeasurementScopeType::USERSPACE_METRIC);
+
+    CounterCollection res;
+
+    if (scope.type == MeasurementScopeType::GROUP_METRIC)
+    {
+        if (group_leader_.is_supported_in(scope.scope))
+        {
+            res.leader = group_leader_;
+            for (auto& ev : group_events_)
+            {
+                if (ev.is_supported_in(scope.scope))
+                {
+                    res.counters.emplace_back(ev);
+                }
+            }
+        }
+    }
+    else
+    {
+        for (auto& ev : userspace_events_)
+        {
+            if (ev.is_supported_in(scope.scope))
+            {
+                res.counters.emplace_back(ev);
+            }
+        }
+    }
```

Comment:
If the CounterCollection knew its scope, this code could live in that class instead and this function collapses to something like this:

```suggestion
    if (scope.type == MeasurementScopeType::GROUP_METRIC)
    {
        return { scope, group_leader_, group_events_ };
    } else {
        return { scope, userspace__events_ };
    }
```

Reply:
I don't know if I'm too dogmatic here, but I personally dont like bringing Scope into CounterCollection as it is designed to be Scope agnostic.

Reply:
If you're setting different things from outside of the class, depending on the scope, it's anything but agnostic. You just don't see it in the CounterCollection. I could add this method to `CounterCollection`, which makes the implicit dependency visible. I'd rather have that relation explicit.

```
MeasurementScopeType scope_type() const {
  return group_leader_ ? MeasurementScopeType::GROUP_METRIC : MeasurementScopeType::USER_METRIC;
}